### PR TITLE
LPS-171006 OutOfMemoryError on startup when there are a lot of very large FragmentEntryLinks

### PR DESCRIPTION
**Pull request by jkappler**
Hi @shuyangzhou, at the end I wasn't able to make the select work, so I worked on your suggestions from the previous PR.

I remembered that we don't support resources in contributed fragments, so the HTML of the contributed fragment and the fragment entry link will always be the same. This is being taken into account in the mass update.

Also, while working on this, it came to my mind that if we add a new column, we won't be able to backport the changes to 7.3. So what I did was use lastPropagationDate column as a sign of the fragment entry link being changed (it doesn't require any upgrade process). If this is null when rendering, the editable fragments are updated, the date is updated and the cache is invalidated.

I tested the changes in MySQL and DB2 and both work fine. I'm also including the tests that Michael and you worked on and adapted an existing one.

Please let us know what you think. If at the end we can backport a column, then I'll make the changes.

Really appreciate your time for helping us on this.

cc:// @ealonso, @ruben-pulido, @lfbesada

### CODE REVIEW

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/service/impl/FragmentEntryLinkLocalServiceImpl.java [hunk 1]
```diff
@@ -562,6 +584,46 @@ public FragmentEntryLink updateFragmentEntryLink(
...
+			preparedStatement.setString(8, configuration);
+			preparedStatement.setString(9, css);
+			preparedStatement.setString(10, html);
+			preparedStatement.setString(11, js);
+			preparedStatement.setInt(12, type);
+
+			preparedStatement.execute();
+		}
+
+		CacheRegistryUtil.clear();
```
Comment (shuyangzhou) [inline modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/service/impl/FragmentEntryLinkLocalServiceImpl.java:624]: @jkappler this is way too overkilled. You should only clear FragmentEntryLink's entity/finder cache.

**Comment by shuyangzhou** [comment]
@jkappler please fix the above issues first, then I need to around of performance analysis for it.

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/service/impl/FragmentEntryLinkLocalServiceImpl.java [hunk 2]
```diff
@@ -562,6 +584,46 @@ public FragmentEntryLink updateFragmentEntryLink(
...
+		}
+
+		StringBundler sb = new StringBundler(4);
+
+		sb.append("update FragmentEntryLink SET configuration = ?, css = ?, ");
+		sb.append("html = ?, js = ?, type_ = ?, lastPropagationDate = ? WHERE ");
+		sb.append("rendererKey = ? AND (configuration != ? OR css != ? OR ");
+		sb.append("html != ? OR js != ? OR type_ != ?)");
+
+		try (Connection connection = DataAccess.getConnection();
```
[the remark this answers is not in the document]
Comment (jkappler) [reply modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/service/impl/FragmentEntryLinkLocalServiceImpl.java]: Hi @shuyangzhou, didn't know about this, thanks! I applied the changes

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/service/impl/FragmentEntryLinkLocalServiceImpl.java (hunk 1 above)
Comment (jkappler) [reply modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/service/impl/FragmentEntryLinkLocalServiceImpl.java]: Thanks @shuyangzhou, changed to 

```
fragmentEntryLinkPersistence.clearCache();

EntityCacheUtil.removeCache(FragmentEntryLink.class.getName()); 
```
Hope it's ok.

**Comment by jkappler** [comment]
@shuyangzhou, changed based on your comments. Re-tested again in mysql and DB2

**Comment by jkappler** [comment]
ci:test:sf

**Comment by jkappler** [comment]
ci:test:relevant

**Comment by shuyangzhou** [comment]
@111anna please run a startup regression check, thanks!

**Comment by jkappler** [comment]
Conflicts fixed

**Comment by jkappler** [comment]
ci:test:echo

**Comment by jkappler** [comment]
running echo tests, jic

cc:// @tim-cao, @david-gutierrez-mesa

**Comment by 111Anna** [comment]
Hi @shuyangzhou, I ran the test on CE and DXP, there seems  no recession.

Here are test results:

CE

Portal Version|GitId|Average Startup (ms)|%|Average Shutdown (ms)|%|Open File|Thread Create|Open Prepared Statement|Open SQL
--|--| --|--|--|--|--|--|--|--|
Baseline| dbfbc84| 22938.3 |- |3286.3|-| 1833 | 101 | 500| 1008 |
Baseline + PR-10928| - |  22901 |0.2%| 3267.6|0.6%| 1833 | 102 | 508 | 1008 |


DXP

Portal Version|GitId|Average Startup (ms)|%|Average Shutdown (ms)|%|Open File|Thread Create|Open Prepared Statement|Open SQL
--|--| --|--|--|--|--|--|--|--|
Baseline| dbfbc84| 25156 |- |3885|-| 1989 | 113| 558| 1097 |
Baseline + PR-10928| - |  25225.3 |-0.3%| 3832.6|1.4%| 1990 | 113 | 604 | 1106 |

**Comment by shuyangzhou** [comment]
See https://github.com/shuyangzhou/liferay-portal/pull/10939

**Pull request closed without merging**